File: plugin/src/HttpLoggingInterceptor.js
Java.perform(function () {
    // 定义日志级别，对应原Java类的Level
    const Level = {
        NONE: 0,
        BASIC: 1,
        HEADERS: 2,
        BODY: 3
    };

    // 配置当前日志级别
    const LOG_LEVEL = Level.BODY;

    // 工具方法：判断是否为明文内容
    function isPlaintext(buffer) {
        try {
            const prefix = Java.use('okio.Buffer').$new();
            const byteCount = buffer.size() < 64 ? buffer.size() : 64;
            buffer.copyTo(prefix, 0, byteCount);

            for (let i = 0; i < 16; i++) {
                if (prefix.exhausted()) break;
                const codePoint = prefix.readUtf8CodePoint();
                if (Java.use('java.lang.Character').isISOControl(codePoint) &&
                    !Java.use('java.lang.Character').isWhitespace(codePoint)) {
                    return false;
                }
            }
            return true;
        } catch (e) {
            return false;
        }
    }

    // 工具方法：判断body是否经过编码
    function bodyEncoded(headers) {
        const contentEncoding = headers.get('Content-Encoding');
        return contentEncoding !== null && contentEncoding.toLowerCase() !== 'identity';
    }

    // 打印日志
    function log(message) {
        console.log(`[HttpLogging] ${message}`);
    }

    // 提取请求信息并打印
    function logRequest(request) {
        if (LOG_LEVEL === Level.NONE) return;

        const method = request.method();
        const url = request.url().toString();
        const hasRequestBody = request.body() !== null;

        // 基本信息
        let logMsg = `--> ${method} ${url}`;
        if (!isLogHeaders() && hasRequestBody) {
            logMsg += ` (${request.body().contentLength()}-byte body)`;
        }
        log(logMsg);

        if (isLogHeaders()) {
            // 打印请求头
            const headers = request.headers();
            for (let i = 0; i < headers.size(); i++) {
                const name = headers.name(i);
                // 跳过已经单独处理的Content-Type和Content-Length
                if (name.toLowerCase() !== 'content-type' && name.toLowerCase() !== 'content-length') {
                    log(`${name}: ${headers.value(i)}`);
                }
            }

            // 打印请求体信息
            if (hasRequestBody) {
                const body = request.body();
                if (body.contentType() !== null) {
                    log(`Content-Type: ${body.contentType()}`);
                }
                if (body.contentLength() !== -1) {
                    log(`Content-Length: ${body.contentLength()}`);
                }

                if (isLogBody() && !bodyEncoded(request.headers())) {
                    try {
                        const buffer = Java.use('okio.Buffer').$new();
                        body.writeTo(buffer);
                        const charset = body.contentType() ? body.contentType().charset(Java.use('java.nio.charset.StandardCharsets').UTF_8) : Java.use('java.nio.charset.StandardCharsets').UTF_8;

                        if (isPlaintext(buffer)) {
                            log('');
                            log(buffer.readString(charset));
                            log(`--> END ${method} (${body.contentLength()}-byte body)`);
                        } else {
                            log(`--> END ${method} (binary ${body.contentLength()}-byte body omitted)`);
                        }
                    } catch (e) {
                        log(`--> END ${method} (error reading body: ${e.message})`);
                    }
                } else {
                    log(`--> END ${method} (encoded body omitted)`);
                }
            } else {
                log(`--> END ${method}`);
            }
        }
    }

    // 提取响应信息并打印
    function logResponse(response, tookMs) {
        if (LOG_LEVEL === Level.NONE) return;

        const code = response.code();
        const message = response.message() || '';
        const url = response.request().url().toString();
        const responseBody = response.body();
        const contentLength = responseBody ? responseBody.contentLength() : -1;
        const bodySize = contentLength !== -1 ? `${contentLength}-byte` : 'unknown-length';

        // 基本响应信息
        let logMsg = `<-- ${code} ${message} ${url} (${tookMs}ms`;
        if (!isLogHeaders()) {
            logMsg += `, ${bodySize} body`;
        }
        logMsg += ')';
        log(logMsg);

        if (isLogHeaders() && responseBody) {
            // 打印响应头
            const headers = response.headers();
            for (let i = 0; i < headers.size(); i++) {
                log(`${headers.name(i)}: ${headers.value(i)}`);
            }

            if (isLogBody() && !bodyEncoded(response.headers())) {
                try {
                    const source = responseBody.source();
                    source.request(Java.use('java.lang.Long').MAX_VALUE); // 缓冲整个响应体
                    const buffer = source.buffer();

                    const charset = responseBody.contentType() ? responseBody.contentType().charset(Java.use('java.nio.charset.StandardCharsets').UTF_8) : Java.use('java.nio.charset.StandardCharsets').UTF_8;

                    if (!isPlaintext(buffer)) {
                        log('');
                        log(`<-- END HTTP (binary ${buffer.size()}-byte body omitted)`);
                        return;
                    }

                    if (contentLength > 0) {
                        log('');
                        log(buffer.clone().readString(charset));
                    }

                    log(`<-- END HTTP (${buffer.size()}-byte body)`);
                } catch (e) {
                    log(`<-- END HTTP (error reading body: ${e.message})`);
                }
            } else {
                log(`<-- END HTTP (encoded body omitted)`);
            }
        } else if (isLogHeaders()) {
            log('<-- END HTTP');
        }
    }

    // 判断是否需要记录头信息
    function isLogHeaders() {
        return LOG_LEVEL === Level.HEADERS || LOG_LEVEL === Level.BODY;
    }

    // 判断是否需要记录体信息
    function isLogBody() {
        return LOG_LEVEL === Level.BODY;
    }

    // Hook OkHttp的Interceptor.Chain的proceed方法
    const Chain = Java.use('okhttp3.Interceptor$Chain');
    Chain.proceed.implementation = function (request) {
        // 记录请求开始时间
        const startNs = Java.use('java.lang.System').nanoTime();

        // 打印请求信息
        logRequest(request);

        try {
            // 执行原方法获取响应
            const response = this.proceed(request);

            // 计算耗时
            const tookMs = Java.use('java.util.concurrent.TimeUnit').NANOSECONDS.toMillis(
                Java.use('java.lang.System').nanoTime() - startNs
            );

            // 打印响应信息
            logResponse(response, tookMs);

            return response;
        } catch (e) {
            log(`<-- HTTP FAILED: ${e.message}`);
            throw e;
        }
    };

    log('HttpLoggingInterceptor hooked successfully');
});

File: packages/hook-item-http/src/HttpLoggingInterceptor.java
package com.kd.http;

import android.util.Log;

import androidx.annotation.NonNull;

import java.io.EOFException;
import java.io.IOException;
import java.nio.charset.Charset;
import java.nio.charset.StandardCharsets;
import java.util.concurrent.TimeUnit;

import okhttp3.Connection;
import okhttp3.Headers;
import okhttp3.Interceptor;
import okhttp3.MediaType;
import okhttp3.Protocol;
import okhttp3.Request;
import okhttp3.RequestBody;
import okhttp3.Response;
import okhttp3.ResponseBody;
import okio.Buffer;
import okio.BufferedSource;

/**
 * 自定义 OkHttp 日志拦截器，支持详细的请求和响应日志记录
 */
public class HttpLoggingInterceptor implements Interceptor {
    private static final Charset UTF8 = StandardCharsets.UTF_8;

    public enum Level {
        NONE,       // 不记录日志
        BASIC,      // 仅记录请求方法、URL、响应状态码和执行时间
        HEADERS,    // 记录基本信息及请求和响应的所有 Headers
        BODY        // 记录基本信息、Headers 及请求和响应的 Body
    }

    private volatile Level level = Level.NONE;

    public HttpLoggingInterceptor() {
    }

    public HttpLoggingInterceptor(Level level) {
        this.level = level;
    }

    public void setLevel(Level level) {
        if (level == null) throw new NullPointerException("level == null. Use Level.NONE instead.");
        this.level = level;
    }

    public Level getLevel() {
        return level;
    }

    @NonNull
    @Override
    public Response intercept(Chain chain) throws IOException {
        Level level = this.level;

        Request request = chain.request();
        if (level == Level.NONE) {
            return chain.proceed(request);
        }

        boolean logBody = (level == Level.BODY);
        boolean logHeaders = (level == Level.BODY || level == Level.HEADERS);

        RequestBody requestBody = request.body();
        boolean hasRequestBody = requestBody != null;

        Connection connection = chain.connection();
        Protocol protocol = connection != null ? connection.protocol() : Protocol.HTTP_1_1;
        String requestStartMessage = "--> " + request.method() + ' ' + request.url() + ' ' + protocol;
        if (!logHeaders && hasRequestBody) {
            requestStartMessage += " (" + requestBody.contentLength() + "-byte body)";
        }
        println(requestStartMessage);

        if (logHeaders) {
            if (hasRequestBody) {
                // Request body headers are only present when installed as a network interceptor.
                // Force them to be included (when available) so there values are known.
                if (requestBody.contentType() != null) {
                    println("Content-Type: " + requestBody.contentType());
                }
                if (requestBody.contentLength() != -1) {
                    println("Content-Length: " + requestBody.contentLength());
                }
            }

            Headers headers = request.headers();
            for (int i = 0, count = headers.size(); i < count; i++) {
                String name = headers.name(i);
                // Skip headers from the request body as they are explicitly logged above.
                if (!"Content-Type".equalsIgnoreCase(name) && !"Content-Length".equalsIgnoreCase(name)) {
                    println(name + ": " + headers.value(i));
                }
            }

            if (!logBody || !hasRequestBody) {
                println("--> END " + request.method());
            } else if (bodyEncoded(request.headers())) {
                println("--> END " + request.method() + " (encoded body omitted)");
            } else {
                Buffer buffer = new Buffer();
                requestBody.writeTo(buffer);

                Charset charset = UTF8;
                MediaType contentType = requestBody.contentType();
                if (contentType != null) {
                    charset = contentType.charset(UTF8);
                }

                println("");
                if (isPlaintext(buffer)) {
                    println(buffer.readString(charset));
                    println("--> END " + request.method() + " (" + requestBody.contentLength() + "-byte body)");
                } else {
                    println("--> END " + request.method() + " (binary " + requestBody.contentLength() + "-byte body omitted)");
                }
            }
        }

        long startNs = System.nanoTime();
        Response response;
        try {
            response = chain.proceed(request);
        } catch (Exception e) {
            println("<-- HTTP FAILED: " + e);
            throw e;
        }
        long tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNs);

        ResponseBody responseBody = response.body();
        long contentLength = responseBody.contentLength();
        String bodySize = contentLength != -1 ? contentLength + "-byte" : "unknown-length";
        println("<-- " + response.code() + (response.message().isEmpty() ? "" : ' ' + response.message())
                + ' ' + response.request().url() + " (" + tookMs + "ms" + (!logHeaders ? ", " + bodySize + " body" : "") + ')');

        if (logHeaders) {
            Headers headers = response.headers();
            for (int i = 0, count = headers.size(); i < count; i++) {
                println(headers.name(i) + ": " + headers.value(i));
            }

            if (!logBody || !responseBodySourceIsReadable(response)) {
                println("<-- END HTTP");
            } else if (bodyEncoded(response.headers())) {
                println("<-- END HTTP (encoded body omitted)");
            } else {
                BufferedSource source = responseBody.source();
                source.request(Long.MAX_VALUE); // Buffer the entire body.
                Buffer buffer = source.buffer();

                Charset charset = UTF8;
                MediaType contentType = responseBody.contentType();
                if (contentType != null) {
                    charset = contentType.charset(UTF8);
                }

                if (!isPlaintext(buffer)) {
                    println("");
                    println("<-- END HTTP (binary " + buffer.size() + "-byte body omitted)");
                    return response;
                }

                if (contentLength > 0) {
                    println("");
                    println(buffer.clone().readString(charset));
                }

                println("<-- END HTTP (" + buffer.size() + "-byte body)");
            }
        }

        return response;
    }

    private boolean responseBodySourceIsReadable(Response response) {
        // Prior to OkHttp 3.9, response.source() returns a BufferSource that
        // is missing the `buffer()` method.
        try {
            ResponseBody responseBody = response.body();
            if (responseBody == null) return false;
            BufferedSource source = responseBody.source();
            if (source == null) return false;
            //noinspection ResultOfMethodCallIgnored
            source.getClass().getMethod("buffer");
            return true;
        } catch (NoSuchMethodException e) {
            return false;
        }
    }

    /**
     * Returns true if the body in question probably contains human readable text. Uses a small sample
     * of code points to detect unicode control characters commonly used in binary file signatures.
     */
    private static boolean isPlaintext(Buffer buffer) {
        try {
            Buffer prefix = new Buffer();
            long byteCount = buffer.size() < 64 ? buffer.size() : 64;
            buffer.copyTo(prefix, 0, byteCount);
            for (int i = 0; i < 16; i++) {
                if (prefix.exhausted()) {
                    break;
                }
                int codePoint = prefix.readUtf8CodePoint();
                if (Character.isISOControl(codePoint) && !Character.isWhitespace(codePoint)) {
                    return false;
                }
            }
            return true;
        } catch (EOFException e) {
            return false; // Truncated UTF-8 sequence.
        }
    }

    private boolean bodyEncoded(Headers headers) {
        String contentEncoding = headers.get("Content-Encoding");
        return contentEncoding != null && !contentEncoding.equalsIgnoreCase("identity");
    }

    private void println(String message) {
        // 这里使用 System.out.println，实际项目中建议使用日志框架如 Logger 或 Logcat
        Log.d("HttpLoggingInterceptor",message);
//        System.out.println(message);
    }
}